File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution(object):
    def topKFrequent(self, nums, k):
        freq = {}
        for i in nums:
            # count frequency 
            # if key exist -> return value
            # else return 0
            freq[i] = freq.get(i, 0) + 1

            # sort the freq
            # according to values not key
            # sort them in decending instead of ascending order
            
        sorted_nums = sorted(freq, key = freq.get, reverse = True)
        # return 1st k elements
        return sorted_nums[:k]
    

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        // 1. Count frequencies

        HashMap<Integer, Integer> map = new HashMap<>();
        for(int n : nums){
            map.put(n, map.getOrDefault(n, 0) + 1);
        }

        // 2. Create Buckets
        List<Integer>[] bucket = new ArrayList[nums.length + 1];
            for(int key : map.keySet()){
                int freq = map.get(key);

                if(bucket[freq] == null){
                    bucket[freq] = new ArrayList<>();
                }

                bucket[freq].add(key);
            }
        
        // 3. Traverse bucket from highest frequency

        int [] ans = new int[k];
        int index = 0;

        for(int i = bucket.length - 1; i >= 0 && index < k; i--){
            if(bucket[i] != null){
                for(int num : bucket[i]){
                    ans[index++] = num;

                    if(index == k){
                        break;
                    }
                }
            }
        }

        return ans;
    }
}